### PR TITLE
[C++][Improvement] Redesign and unify the implementation of validation in C++ Writer/Builder

### PR DESCRIPTION
**Pull request by lixueclaire**
## Proposed changes

We need to develop a more comprehensive and consistent approach to validation across all of GraphAr's tools and libraries. As the first step, this change redesigned and unified the implementation of validation for `Writer` and `Builder` in the C++ library.

- clarify and define different validation levels in the utils
- hide the `Validate` interface to users and use `validate_level` as a parameter of the writing operations
- simplify and unify the implementation for validation in `Writer` and `Builder`
- add a `Clear()` method to high-level `Builder`, enabling to add data and then dump for multiple rounds
- add test to prove the interface changes
- update C++ API Reference to include the `ValidateLevel` type

## Types of changes

What types of changes does your code introduce to GraphAr?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/alibaba/GraphAr/blob/main/CONTRIBUTING.rst) doc
- [x] I have signed the CLA
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

Solve #183

### CODE REVIEW

File: cpp/src/arrow_chunk_writer.cc [hunk 1]
```diff
@@ -90,7 +90,42 @@ Result<std::shared_ptr<arrow::Table>> ExecutePlanAndCollectAsTable(
...
+                                      ValidateLevel validate_level) const
+    noexcept {
+  // use the writer's validate level
+  if (validate_level == ValidateLevel::default_validate)
+    validate_level = validate_level_;
+  // no validate
+  if (validate_level == ValidateLevel::no_validate)
+    return Status::OK();
+  // weak & strong validate
+  if (count < 0)
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:103]: please follow the [google c++ code style](https://google.github.io/styleguide/cppguide.html#Formatting_Looping_Branching), if body can't not fit on one line, use braces

File: cpp/src/arrow_chunk_writer.cc [hunk 2]
```diff
@@ -90,7 +90,42 @@ Result<std::shared_ptr<arrow::Table>> ExecutePlanAndCollectAsTable(
...
+    noexcept {
+  // use the writer's validate level
+  if (validate_level == ValidateLevel::default_validate)
+    validate_level = validate_level_;
+  // no validate
+  if (validate_level == ValidateLevel::no_validate)
+    return Status::OK();
+  // weak & strong validate
+  if (!vertex_info_.ContainPropertyGroup(property_group))
+    return Status::InvalidOperation(
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:121]: Ditto

File: cpp/src/arrow_chunk_writer.cc [hunk 3]
```diff
@@ -100,17 +135,14 @@ Status VertexPropertyWriter::Validate(
   // no validate
   if (validate_level == ValidateLevel::no_validate)
     return Status::OK();
-  // weak validate
+  // validate property_group & chunk_index
+  GAR_RETURN_NOT_OK(validate(property_group, chunk_index, validate_level));
+  // weak validate for the input_table
   if (input_table->num_rows() > vertex_info_.GetChunkSize())
     return Status::OutOfRange(
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:142]: Ditto

File: cpp/src/arrow_chunk_writer.cc [hunk 4]
```diff
@@ -174,58 +212,95 @@ Status VertexPropertyWriter::WriteChunk(
...
   // use the writer's validate level
   if (validate_level == ValidateLevel::default_validate)
     validate_level = validate_level_;
   // no validate
   if (validate_level == ValidateLevel::no_validate)
     return Status::OK();
-  // weak validate
+  // weak & strong validate for adj list type
   if (!edge_info_.ContainAdjList(adj_list_type_))
     return Status::InvalidOperation(
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:264]: Ditto

File: cpp/src/arrow_chunk_writer.cc [hunk 5]
```diff
@@ -174,58 +212,95 @@ Status VertexPropertyWriter::WriteChunk(
...
+  if (validate_level == ValidateLevel::default_validate)
+    validate_level = validate_level_;
+  // no validate
+  if (validate_level == ValidateLevel::no_validate)
+    return Status::OK();
+  // validate for adj list type & index
+  GAR_RETURN_NOT_OK(validate(vertex_chunk_index, chunk_index, validate_level));
+  // weak & strong validate for property group
+  if (!edge_info_.ContainPropertyGroup(property_group, adj_list_type_))
+    return Status::InvalidOperation(
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:287]: Ditto

File: cpp/src/arrow_chunk_writer.cc [hunk 6]
```diff
@@ -90,7 +90,42 @@ Result<std::shared_ptr<arrow::Table>> ExecutePlanAndCollectAsTable(
 
 // implementations for VertexPropertyChunkWriter
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:92]: maybe to add a comment to indicate the validate method for which operation in .cc?

**Review by acezen — COMMENTED** [review]
Thanks for the job, it's overall look good to me. there are some places I think we can revise:
- In your design, the `default_validate`  is invalid to writer/builder's validate level, so we need to note this in comment to users and add a check if validate_level is `default_validate` or not in constructors.
- It seems that weak/strong's check is different for operation. Beside the brief doc for `ValidateLevel`, how about add some docs to the operations that what will `weak` and `strong` check for the operation? I think that would help develop and user to understand the operation's behavior.
- There are some codes not follow the google c++ code style, please fix them.

**Comment by lixueclaire** [comment]
> Thanks for the job, it's overall look good to me. there are some places I think we can revise:
> 
> * In your design, the `default_validate`  is invalid to writer/builder's validate level, so we need to note this in comment to users and add a check if validate_level is `default_validate` or not in constructors.
> * It seems that weak/strong's check is different for different operations. Beside the brief doc for `ValidateLevel`, how about add some docs to the operations that what will `weak` and `strong` check for the operation? I think that would help develop and user to understand the operation's behavior.
> * There are some codes not follow the google c++ code style, please fix them.

Thanks for your suggestions. For the second comment, weak/strong check is consistent for different operations now, strong_validate will also check the schema, while weak_validate only validates index, count, adj_list type, property group and size.

**Comment by acezen** [comment]
The code seems to meet the [clang-format config](https://github.com/alibaba/GraphAr/blob/main/cpp/.clang-format), no need to refine.

File: cpp/src/arrow_chunk_writer.cc (hunk 6 above)
Comment (lixueclaire) [reply cpp/src/arrow_chunk_writer.cc:92]: sovled

**Comment by lixueclaire** [comment]
I have checked the `validate_level` passed to constructors; added comments in src to explain the implementations for validations added docstrings for different validate levels for each `Writer`/`Builder`; and fixed the code style. @acezen, could you please have a look again?

**Review by acezen — APPROVED** [review]
LGTM

**Pull request merged**